### PR TITLE
Support for model outlining

### PR DESCRIPTION
**Pull request by jasonbeverage**
Hi all,

I've done a bit of work adding support to Cesium to provide a halo effect for outlining models when they are selected (or any other reason you want to highlight them).

I've included a new sandcastle example that you can try here:

https://cesiumdemos.netlify.com/apps/sandcastle/?src=development/3D%20Models%20Outline.html&label=Development

Most of the modifications were to Model.js.  It provides a two pass rendering approach using the stencil buffer, where the model is drawn normally first, filling the stencil buffer, then the model is drawn again using a shader that extrudes the normals and only renders where the original model didn't render.

This is my first time digging around the guts of the model rendering code in Cesium, so I'd appreciate any feedback on the approach I took.

Thanks!

Jason

@lilleyse TODO:
* [x] To support multiple models simultaneously the stencil values need to be cleared between commands. I have been having trouble thinking of a way to clear these values during the silhouette color pass, so most likely I will go with a clear command of the screen-space region of the model according to @pjcozzi's suggestion. This will require a few modifications including supporting a `RenderState` in `ClearCommand`. This approach sort of breaks down for models that contain both opaque and transparent commands since the stencil buffer will get trashed in-between the opaque and translucent passes. All commands could be treated as translucent, but if OIT isn't supported the commands will be sorted and the clear commands won't. Essentially what's required for this all to work is render all model commands at the same time, opaque and translucent, immediately followed by the clear command - but a paradigm like list doesn't really exist in the rendering engine. Alternatively, if only the extruded back faces are rendered then the stencil writes can be done in the silhouette color pass and no clear command is needed. At this point though, it's useless to use stenciling at all and the artifacts are worse as @jasonbeverage noted. One further solution is to support only 255 silhouettes at a time - but this is limiting and has some edge cases in terms of constantly updating render states. **Decided to go with this last approach and accept the minimal and somewhat unlikely artifacts**
* [x] If going with the clear command approach, determine the screen space region of the model and apply a scissor test for that. **Not applicable**
* [x] The shader references `v_normal` and assumes the normal is in view space, both of which are true for Cesium's sample models but not for all models.
* [ ] More unit tests, especially interactions with `color` with translucency.
* [x] Update to `CHANGES.md`.
* [ ] Update CZML documentation.

### CODE REVIEW

**Comment by hpinkos** [comment]
Wow, the demo looks awesome @jasonbeverage!  Thanks so much for submitting this.

Do we have a contributor license agreement (CLA) from you?  If not, can you please send one over so we can review this? You can find instructions in [CONTRIBUTING.md](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/CONTRIBUTING.md#contributor-license-agreement-cla)

**Comment by jasonbeverage** [comment]
Yup you should have one on file, I've had other PR's merged.  Thanks!

**Comment by hpinkos** [comment]
Great, thanks @jasonbeverage!

You have some jsHint errors.  And I think we'll probably want some unit tests for this.  See the [pull request guidelines](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/CONTRIBUTING.md#pull-request-guidelines) in our contributing documentation. 

We're a little busy around here this week and we have some people on vacation, so we might not be able to review this right away.  I think this would be a really useful feature though, and I'd love to see it merged soon =)

**Comment by mramato** [comment]
Just wanted to say that this is awesome.  I was talking offline with @bagnell and he thought the same technique could be extended to other primitives (having a generic way to highlight all primitives is definitely something we've wanted for a long time).  Our selection icon is good for HTML, but having a highlight that scales with the object is better.  While I don't expect it to be part of this PR, having a new issue to track adding similar capabilities to geometry/labels/billboards/points would be great.

While I'm not qualified to review and merge this, from a functionality standpoint my one comment would be that we map the width parameter to use screen space units so that it's easier for users to understand.

Thanks @jasonbeverage.

**Comment by jasonbeverage** [comment]
Glad you like it @mramato :)  It seems like it would be easy to extend to other graphics like labels and billboards, if we did that perhaps it would make sense to stick the highlight properties on the entity object itself instead of on the model, billboard, etc.

I agree on the screen space mapping instead of the weird clip space setting I have now, let me see what I can do.

**Comment by jasonbeverage** [comment]
I've pushed some changes that should fix the jsHint issues and also changed the highlightSize to be in pixels instead of clipspace.

I have a question about how the uniforms work in Cesium though.  I added a new function called getUniformNameForSemantic that tries to get a uniform from GLTF based on the semantic.  For the projection matrix I had no fallback to czm_projection if I couldn't find the projection matrix, but when I tried to get the VIEWPORT semantic I was getting no uniform name back (I assume b/c the models don't use VIEWPORT in their shaders) so I fell back to czm_viewport.  If the uniform wasn't found I figured I needed to define it in the shader b/c it wasn't used, but when I tried to append it in createHighlightVertexShaderSource using something like this:
if (!viewportUniformDefined) {
   highlightMain +=  "uniform vec4 " + viewportName + ";\n"
}

However, doing that I get get a czm_viewport already defined shader compiler error even though czm_viewport actually isn't defined in the model's shader.  I can apparently use czm_viewport just fine without defining it in my shader, is the model's shader eventually merged with some uber shader down the line?

Thanks!

**Comment by lilleyse** [comment]
> I can apparently use czm_viewport just fine without defining it in my shader, is the model's shader eventually merged with some uber shader down the line?

Yup, before the shader is finalized it checks for any of the cesium built-in uniforms like `czm_viewport` and appends the uniform declarations to the shader: https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Renderer/ShaderSource.js#L60

**Comment by jasonbeverage** [comment]
Oh great, so that means that I actually don't need to get the semantic name for projection and viewport, I can just use czm_projection and czm_viewport freely right?

**Comment by lilleyse** [comment]
Yeah that should be doable.

**Comment by jasonbeverage** [comment]
Great, I just pushed a change doing that and it worked fine.  Thanks!

File: Source/Scene/Model.js [hunk 1]
```diff
@@ -2972,11 +3069,58 @@ define([
...
                     uniformMap : uniformMap,
-                    renderState : rs,
+                    renderState : drawRS,
                     owner : owner,
                     pass : isTranslucent ? Pass.TRANSLUCENT : Pass.OPAQUE
                 });
 
+                // Setup the stencil command for the highlight
+                var highlightRS = clone(rs);
+                highlightRS.stencilTest = {
```
Comment (pjcozzi) [inline Source/Scene/Model.js]: Ah, actually this isn't culling at all.  Why?

File: Source/Scene/Model.js [hunk 2]
```diff
@@ -2972,11 +3069,58 @@ define([
...
+                    },
+                    backOperation : {
+                        fail : WebGLConstants.KEEP,
+                        zFail : WebGLConstants.KEEP,
+                        zPass : WebGLConstants.REPLACE
+                    }
+                };
+                highlightRS.cull = {
+                    enabled : false,
+                    face : WebGLConstants.FRONT
```
Comment (pjcozzi) [inline Source/Scene/Model.js]: Setting `face` isn't needed here.

File: Source/Scene/Model.js [hunk 3]
```diff
@@ -2972,11 +3069,58 @@ define([
...
+                };
+                highlightRS.depthTest = false;
+
+                highlightRS = RenderState.fromCache(highlightRS);
+
+                // Setup the highlight color and size uniforms.
+                uniformMap.u_highlightColor = createHighlightColorFunction(model);
+                uniformMap.u_highlightSize = createHighlightSizeFunction(model);
+
+                var highlightCommand = new DrawCommand({
```
Comment (pjcozzi) [inline Source/Scene/Model.js]: Can we create these commands and related resources on-demand the first time `highlight` is `true`?

File: Source/Scene/Model.js [hunk 4]
```diff
@@ -2972,11 +3069,58 @@ define([
...
+                        count : count,
+                        offset : offset,
+                        shaderProgram : rendererhighlightPrograms[technique.program],
+                        uniformMap : uniformMap,
+                        renderState : highlightRS,
+                        owner : owner,
+                        pass : isTranslucent ? Pass.TRANSLUCENT : Pass.OPAQUE
+                });
+
+
```
Comment (pjcozzi) [inline Source/Scene/Model.js]: Extra whitespace.

File: Source/Scene/Model.js [hunk 5]
```diff
@@ -3037,7 +3186,10 @@ define([
                     command : command,
                     pickCommand : pickCommand,
                     command2D : command2D,
-                    pickCommand2D : pickCommand2D
+                    pickCommand2D : pickCommand2D,
+                    highlightCommand: highlightCommand,
+                    highlightCommand2D: highlightCommand2D
+
```
Comment (pjcozzi) [inline Source/Scene/Model.js]: Whitespace.

File: Source/Scene/Model.js [hunk 6]
```diff
@@ -3234,6 +3386,11 @@ define([
                                 BoundingSphere.clone(command.boundingVolume, pickCommand.boundingVolume);
                             }
 
+                            var highlightCommand = primitiveCommand.highlightCommand;
+                            Matrix4.clone(command.modelMatrix, highlightCommand.modelMatrix);
+                            BoundingSphere.clone(command.boundingVolume, highlightCommand.boundingVolume);
+
+
```
Comment (pjcozzi) [inline Source/Scene/Model.js]: Whitespace.

File: Source/Scene/Model.js [hunk 7]
```diff
@@ -2972,11 +3069,58 @@ define([
...
                     receiveShadows : receiveShadows,
                     uniformMap : uniformMap,
-                    renderState : rs,
+                    renderState : drawRS,
                     owner : owner,
                     pass : isTranslucent ? Pass.TRANSLUCENT : Pass.OPAQUE
                 });
 
+                // Setup the stencil command for the highlight
+                var highlightRS = clone(rs);
```
Comment (pjcozzi) [inline Source/Scene/Model.js]: Shouldn't the blend state for this depend on the alpha value of `hightlightColor`?  We might also be able to get always with just always using alpha blending for this pass.

**Comment by pjcozzi** [comment]
@jasonbeverage cool stuff, I'm sure users will appreciate this and will get a kick out of animating the highlight size and color/translucency!

Comments:

* Have you tried multiple overlapping 3D models?  3D models partly behind terrain?  Given that the state of the stencil buffer is not known before each model, I think we are going to run into issues here.  We could, of course, scissor around the model and clear the stencil before each model, but I suspect there is a clearer and faster approach.
* Maybe some of our code comments are out of date, but I am actually surprised that this works at all since `getContext` does not create a stencil buffer by default.  I am fine with explicitly defaulting `stencil : true`, but the spec states that when this is `false`, it should not accidentally behave like there is one.  Maybe I missed it since I only looked in GitHub.
   * Either way, the number of stencil bits may be zero so we should give users the ability to know if highlighting will work, e.g., a supported property that checks https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Renderer/Context.js#L512.  We should also not render the highlight pass if it won't work.
* Add unit tests, see https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Documentation/Contributors/TestingGuide/README.md
* Update CHANGES.md

Let me know if you have any questions on my code comments.  They were all minor except the lazy creation of renderer resources, which will also be easy, but is important so users don't pay for things they don't use.

Also, we planned to implement this with an upcoming post-processing framework; the screen-space technique will likely be much faster, especially given the typical CPU bottlenecks in Cesium apps, but I don't know how the quality will compare; I suspect it will be good.  No concern now, just a heads up that we might significantly change this later...and you are more than welcome to help!

**Comment by jasonbeverage** [comment]
Thanks for the feedback Patrick, I'll update the PR here soon and try to address your comments.  Can you point me to some example unit tests that be applicable in this situation that I could try to draw inspiration from?

**Comment by pjcozzi** [comment]
For unit tests, we often rendering into a 1x1 viewport and check the color; however, even this is not super reliable across browsers, drivers, etc. so we have been doing more tests that check what commands were created (in this case, for example, you could validate the stencil render state).

For simple property set/get, see https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Specs/Scene/ModelSpec.js#L214-L233

For command testing, see https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Specs/Scene/PrimitiveSpec.js#L447-L467

BTW, we will likely change this to use "derived commands" so that the renderer can automatically generate highlight commands for any type of primitive instead of require each primitive to have custom code to create these commands.  It totally isn't required for this PR, but if it is something you are interested in, @bagnell and I could advise.

**Comment by pjcozzi** [comment]
@jasonbeverage any update here?  We will have some time to do another review at the bug bash next week.  We could also just finish this off for you (maybe even at the bug bash) if you prefer.

**Comment by jasonbeverage** [comment]
Hey Patrick, let me address some of your comments on Monday, if there is
anything left after that you're more than welcome to make any changes you
want to get it merged.  Thanks!

On Sat, Oct 15, 2016, 11:45 AM Patrick Cozzi <notifications@github.com>
wrote:

@jasonbeverage <https://github.com/jasonbeverage> any update here? We will
have some time to do another review at the bug bash next week. We could
also just finish this off for you (maybe even at the bug bash) if you
prefer.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/AnalyticalGraphicsInc/cesium/pull/4314#issuecomment-253992092>,
or mute the thread

**Comment by pjcozzi** [comment]
Sounds great, thanks again for this nice contribution!

File: Apps/Sandcastle/gallery/development/3D Models Outline.html [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+
+    entity = viewer.entities.add({
+        name : url,
+        position : position,
+        orientation : orientation,
+        model : {
+            uri : url,
+            minimumPixelSize : 128,
+            maximumScale : 20000,
+            highlight: viewModel.highlight,
```
Comment (jasonbeverage) [inline Apps/Sandcastle/gallery/development/3D Models Outline.html]: Done

File: Source/DataSources/ModelGraphics.js [hunk 9]
```diff
@@ -51,6 +51,9 @@ define([
      * @param {Property} [options.receiveShadows=true] Deprecated, use options.shadows instead. A boolean Property specifying whether the model receives shadows from shadow casters in the scene.
      * @param {Property} [options.shadows=ShadowMode.ENABLED] An enum Property specifying whether the model casts or receives shadows from each light source.
      * @param {Property} [options.heightReference=HeightReference.NONE] A Property specifying what the height is relative to.
+     * @param {Property} [options.highlight=false] Whether to highlight the model using an outline
+     * @param {Property} [options.highlightColor=Color(1.0, 0.0, 0.0, 1.0)] The highlight color for the outline.
```
Comment (jasonbeverage) [inline Source/DataSources/ModelGraphics.js]: Done.

File: Source/Scene/Model.js [hunk 10]
```diff
@@ -400,6 +402,34 @@ define([
...
+         * @default false
+         */
+        this.highlight = defaultValue(options.highlight, false);
+
+        /**
+         * The highlight color
+         *
+         * @type {Cartesian4}
+         *
+         * @default Color(1.0, 0.0, 0.0, 1.0)
```
Comment (jasonbeverage) [inline Source/Scene/Model.js]: Done.

File: Source/Scene/Model.js [hunk 11]
```diff
@@ -400,6 +402,34 @@ define([
...
+        /**
+         * The size of the highlight
+         *
+         * @type {Float}
+         *
+         * @default 0.002
+         */
+        this.highlightSize = 0.002;
+
+
```
Comment (jasonbeverage) [inline Source/Scene/Model.js]: Done.

File: Source/Scene/Model.js [hunk 12]
```diff
@@ -2768,6 +2830,7 @@ define([
                 u.jointMatrixUniformName = jointMatrixUniformName;
             }
         }
+
```
Comment (jasonbeverage) [inline Source/Scene/Model.js]: Done.

File: Source/Scene/Model.js (hunk 5 above)
Comment (jasonbeverage) [inline Source/Scene/Model.js]: done

File: Source/Scene/Model.js (hunk 6 above)
Comment (jasonbeverage) [inline Source/Scene/Model.js]: done.

File: Source/Scene/Model.js (hunk 7 above)
Comment (jasonbeverage) [inline Source/Scene/Model.js]: Good catch, fixed and updated the Sandcastle demo to have an opacity slider, it looks nice with a partially transparent outline.

File: Source/Scene/Model.js (hunk 1 above)
Comment (jasonbeverage) [inline Source/Scene/Model.js]: I agree, we can look at modifying the shader later to discard out parts with alpha = zero, but perhaps better to wait until it's really necessary and we have a model that really exhibits the problem.

I originally had this culling out front faces like you mentioned, which is why I had the cull face set to WebGLConstants.FRONT.  It works, but didn't look as nice on most of the test models as just disabling cull face altogether.  You start to get gaps and breaks in the outline really fast as you increase the size.

I believe the issue is that the shell expansion techniques are assuming that you are dealing with a continuous closed mesh that will extrude as a whole, instead of dealing with individual nodes in the models that are disconnected and extrude as separate parts.

File: Source/Scene/Model.js (hunk 3 above)
Comment (jasonbeverage) [inline Source/Scene/Model.js]: I tried to mimic what was done with the pickCommands, so kept the highlight command and resource creation close to where the pick commands are generated.  If you don't mind I'll defer this to you to move around so I don't break anything :)

**Comment by jasonbeverage** [comment]
Quick comment on the stencil test, you're right that it should be disabled by default, but at least in Chrome on Windows there is no difference between enabling it and disabling it.

I know the stenciling is working because if I remove the stencilTest code in Model.js you can only see the highlight color and not the actual model.

Interestingly enough context.stencilBits will return 0 if you don't request a stencil buffer, and 8 if you do, but I can't see any difference in actual rendering behavior. 

I've updated the 3D Models Outline example to request a stencil buffer like this:

```
var viewer = new Cesium.Viewer('cesiumContainer', {
    infoBox : false,
    selectionIndicator : false,
    shadows : true,
    // Explicitly request a stencil buffer so highlighting will work.
    contextOptions: {
        webgl : {
            stencil: true
        }
    }
});
```

**Comment by jasonbeverage** [comment]
@pjcozzi I've pushed changes that should address most of your code review and I've added some unit tests as well.  I've changed the sandcastle demo a little to have an opacity slider as well to test the alpha blending.  New demo is uploaded here:
https://cesiummodelhighlight.netlify.com/apps/sandcastle/?src=development/3D%20Models%20Outline.html&label=Development

I'll turn it over to you now unless you need anything else from me, thanks!

**Comment by pjcozzi** [comment]
Thanks for all the updates @jasonbeverage!

Sorry we didn't get to this at the bug bash; however, we will have a code sprint in early December, and either @bagnell or I will finish this off so it will ship in the Cesium 1.29 release in January.  So you won't have to maintain this branch too much longer.

**Comment by jasonbeverage** [comment]
That sounds great @pjcozzi , thanks for working with me to get it merged!

**Comment by pjcozzi** [comment]
@jasonbeverage we are going to finish this up next week at our code sprint so this should be in Cesium 1.29 in January.  I'm going to do another review now to prep for next week.  Don't worry about making any additional changes; most likely @lilleyse will do any needed changs.  Thanks again for this!

File: Apps/Sandcastle/gallery/development/3D Models Outline.html [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+    'use strict';
+//Sandcastle_Begin
+var viewer = new Cesium.Viewer('cesiumContainer', {
+    infoBox : false,
+    selectionIndicator : false,
+    shadows : true,
+    // Explicitly request a stencil buffer so highlighting will work.
+    contextOptions: {
+        webgl : {
+            stencil: true
```
Comment (pjcozzi) [inline Apps/Sandcastle/gallery/development/3D Models Outline.html]: @bagnell is this needed since polygons on terrain also require stencil?  Last time I reviewed this, it was not clear to me where Cesium requested stencil.  Seems like it should default to `true` now.

File: Apps/Sandcastle/gallery/development/3D Models Outline.html [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+    entity = viewer.entities.add({
+        name : url,
+        position : position,
+        orientation : orientation,
+        model : {
+            uri : url,
+            minimumPixelSize : 128,
+            maximumScale : 20000,
+            highlight: viewModel.highlight,
+            highlightSize: viewModel.highlightSize,
```
Comment (pjcozzi) [inline Apps/Sandcastle/gallery/development/3D Models Outline.html]: Throughout, we should consider calling this "silhouette", instead of "highlight", since "hightlight" could mean many things like #4547 or highlighting 3D Tiles features.

File: Apps/Sandcastle/gallery/development/3D Models Outline.html [hunk 15]
```diff
@@ -0,0 +1,184 @@
+<!DOCTYPE html>
```
Comment (pjcozzi) [inline Apps/Sandcastle/gallery/development/3D Models Outline.html]: @lilleyse it could be cool to combine this example with #4547 and then fade in/out both coloring the model and exlarging the silhouette over time.  Tweens are not part of the [public API](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Scene/TweenCollection.js) so we might want to expose them if time permits.

File: Apps/Sandcastle/gallery/development/3D Models.html [hunk 16]
```diff
@@ -22,7 +22,9 @@
 </style>
 <div id="cesiumContainer" class="fullSize"></div>
 <div id="loadingOverlay"><h1>Loading...</h1></div>
-<div id="toolbar"></div>
+<div id="toolbar">
```
Comment (pjcozzi) [inline Apps/Sandcastle/gallery/development/3D Models.html]: Undo this change.

File: Source/Scene/Model.js [hunk 17]
```diff
@@ -399,6 +403,33 @@ define([
...
+        this.highlight = defaultValue(options.highlight, false);
+
+        /**
+         * The highlight color
+         *
+         * @type {Color}
+         *
+         * @default Color()
+         */
+        this.highlightColor = defaultValue(options.highlightColor, new Color());
```
Comment (pjcozzi) [inline Source/Scene/Model.js]: Minor, but use the ternary operator to avoid the allocation when it is not needed.

File: Source/Scene/Model.js [hunk 18]
```diff
@@ -1724,6 +1756,23 @@ define([
...
     }
 
+    function createHighlightVertexShaderSource(vertexShaderSource) {
+        var renamedVS = ShaderSource.replaceMain(vertexShaderSource, 'czm_old_main');
+        // Modified from http://forum.unity3d.com/threads/toon-outline-but-with-diffuse-surface.24668/
+        var highlightMain = 'uniform float u_highlightSize;\n' +
+            'void main() \n' +
+            '{ \n' +
+            '  czm_old_main(); \n' +
+            '  vec3 n = normalize(v_normal);\n' +
```
Comment (pjcozzi) [inline Source/Scene/Model.js]: Is `v_normal` already normalized?

Is it safe to assume `v_normal` exists?  We likely need to get the name from the uniform semantic and then do the model-view transform here.

File: Source/Scene/Model.js [hunk 19]
```diff
@@ -2931,7 +3008,28 @@ define([
 
                 var castShadows = ShadowMode.castShadows(model._shadows);
                 var receiveShadows = ShadowMode.receiveShadows(model._shadows);
-                
+
+                // Setup the stencil for the first pass
+                var drawRS = clone(rs);
+                drawRS.stencilTest = {
```
Comment (pjcozzi) [inline Source/Scene/Model.js]: Here and below remove the [defaults](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Renderer/RenderState.js#L151-L166) unless it makes it too hard to follow.

File: Source/Scene/Model.js [hunk 20]
```diff
@@ -2944,7 +3042,53 @@ define([
                     castShadows : castShadows,
                     receiveShadows : receiveShadows,
                     uniformMap : uniformMap,
-                    renderState : rs,
+                    renderState : drawRS,
```
Comment (pjcozzi) [inline Source/Scene/Model.js]: Let's only use this render state when highlight is enabled.

File: Source/Scene/Model.js [hunk 21]
```diff
@@ -2944,7 +3042,53 @@ define([
...
+                    backOperation : {
+                        fail : WebGLConstants.KEEP,
+                        zFail : WebGLConstants.KEEP,
+                        zPass : WebGLConstants.REPLACE
+                    }
+                };
+                highlightRS.cull = {
+                    enabled : false
+                };
+                highlightRS.depthTest = false;
```
Comment (pjcozzi) [inline Source/Scene/Model.js]: This and the blending below should only be set when the highlight color alpha is not one.

File: Source/Scene/Model.js (hunk 3 above)
Comment (pjcozzi) [inline Source/Scene/Model.js]: @lilleyse let me know if you have any questions on this one; I prefer creating these on-demand since they may not be used in many apps, and users should only pay for what they use when it's reasonable for us to implement.

File: Source/Scene/Model.js [hunk 22]
```diff
@@ -2944,7 +3042,53 @@ define([
...
+                highlightRS.stencilTest = {
+                    enabled : true,
+                    frontFunction : WebGLConstants.NOTEQUAL,
+                    backFunction : WebGLConstants.NOTEQUAL,
+                    reference : 1,
+                    mask : ~0,
+                    frontOperation : {
+                        fail : WebGLConstants.KEEP,
+                        zFail : WebGLConstants.KEEP,
+                        zPass : WebGLConstants.REPLACE
```
Comment (pjcozzi) [inline Source/Scene/Model.js]: I think here and `backOperation.zPass` could just be `KEEP`.

File: Source/Scene/Model.js [hunk 23]
```diff
@@ -3808,19 +3967,42 @@ define([
...
+                                boundingVolume = nc.command.boundingVolume;
+                                if (frameState.mode === SceneMode.SCENE2D &&
+                                    (boundingVolume.center.y + boundingVolume.radius > idl2D || boundingVolume.center.y - boundingVolume.radius < idl2D)) {
+                                    commandList.push(nc.highlightCommand2D);
+                                }
+                            }
+                        }
+                    }
+                    else {
+                        console.log("Model highlighting not supported, stencilBits = " + context.stencilBits + ".  Request a stencil buffer when initializing the Viewer");
```
Comment (pjcozzi) [inline Source/Scene/Model.js]: Instead of this `console.log`, let's just throw an exception when highlight is enabled and this is not supported; mirror how polygons on terrain work.

File: Specs/Scene/ModelSpec.js [hunk 24]
```diff
@@ -10,6 +10,7 @@ defineSuite([
         'Core/defaultValue',
         'Core/defined',
         'Core/defineProperties',
+        'Core/Color',
```
Comment (pjcozzi) [inline Specs/Scene/ModelSpec.js]: @lilleyse this will need a lot more unit tests.

File: Apps/Sandcastle/gallery/development/3D Models Outline.html (hunk 13 above)
Comment (pjcozzi) [inline Apps/Sandcastle/gallery/development/3D Models Outline.html]: Also see https://github.com/AnalyticalGraphicsInc/cesium/pull/4314#issuecomment-254303838

File: Specs/Scene/ModelSpec.js (hunk 24 above)
Comment (pjcozzi) [inline Specs/Scene/ModelSpec.js]: Related: https://github.com/AnalyticalGraphicsInc/cesium/pull/4314#issuecomment-249663870

**Comment by pjcozzi** [comment]
@lilleyse echoing some of my comments here, https://github.com/AnalyticalGraphicsInc/cesium/pull/4314#issuecomment-249388669, and here, https://github.com/AnalyticalGraphicsInc/cesium/pull/4314#issuecomment-249663870:
* We will either need to render another pass to clear the stencil buffer or scissor out the projected bounding sphere and clear the stencil.  The later is likely much more efficient and @likangning93 can advise on the projection code.
* I would get this fully cleaned up and then consider if we want to tackle this with derived commands, which would allow supporting this everywhere...think through how this would work with batching, e.g., highlighting one feature in 3D Tiles.

**Comment by pjcozzi** [comment]
Just those comments, the scope here is within the code sprint.

**Comment by jasonbeverage** [comment]
Hey @pjcozzi , thanks again for finishing this up and merging it!  Happy to see my first non trivial PR get merged into Cesium 😄

**Comment by lilleyse** [comment]
Reorganized a bit. There are still some issue I need to figure out:

* [ ] To support multiple models simultaneously the stencil values need to be cleared between commands. I have been having trouble thinking of a way to clear these values during the silhouette color pass, so most likely I will go with a clear command of the screen-space region of the model according to @pjcozzi's suggestion. This will require a few modifications including supporting a `RenderState` in `ClearCommand`. This approach sort of breaks down for models that contain both opaque and transparent commands since the stencil buffer will get trashed in-between the opaque and translucent passes. All commands could be treated as translucent, but if OIT isn't supported the commands will be sorted and the clear commands won't. Essentially what's required for this all to work is render all model commands at the same time, opaque and translucent, immediately followed by the clear command - but a paradigm like list doesn't really exist in the rendering engine. Alternatively, if only the extruded back faces are rendered then the stencil writes can be done in the silhouette color pass and no clear command is needed. At this point though, it's useless to use stenciling at all and the artifacts are worse as @jasonbeverage noted. One further solution is to support only 255 silhouettes at a time - but this is limiting and has some edge cases in terms of constantly updating render states.
* [ ] If going with the clear command approach, determine the screen space region of the model and apply a scissor test for that.
* [ ] The shader references `v_normal` and assumes the normal is in view space, both of which are true for Cesium's sample models but not for all models.
* [ ] More unit tests, especially interactions with `color` with translucency.
* [ ] Update to `CHANGES.md`.
* [ ] Update CZML documentation.

**Comment by lilleyse** [comment]
This is updated now.

There were too many special cases with the stencil-clearing approach so I decided to instead go with using the unique stencil value per model approach. 256 models can be outlined without any artifacts, but once it goes over there would be some minor stencil artifacts if two models with the same stencil id happen to overlap. I think this is a rare situation and a reasonable trade-off. Another plus is there are at most 2 render passes, whereas the other approach often required three if translucency was involved.

This PR should work fine when OIT is supported. When it's not, the sorting causes stencil commands to be rendered in arbitrary order. A possible solution here is to create a derived command that combines multiple commands and always renders them together. All the model commands would be rendered sequentially, followed by all the stencil commands. Does this approach sound ok?

So while this PR is not complete yet it's worth a look.

**Comment by lilleyse** [comment]
> A possible solution here is to create a derived command that combines multiple commands and always renders them together. All the model commands would be rendered sequentially, followed by all the stencil commands. Does this approach sound ok?

Another idea for the non-OIT situation: have a separate render pass right after translucency. The downside is translucent silhouettes will not be rendered in the correct sorted order.

**Comment by pjcozzi** [comment]
> Another idea for the non-OIT situation: have a separate render pass right after translucency. The stencil buffer will be completely ready. The downside is translucent silhouettes will not be rendered in the correct sorted order.

Given that this is a short/medium-term solution for silhouettes that will be replaced by a screen-space approach, I would not do this architecture change.

**Comment by pjcozzi** [comment]
> 256 models can be outlined without any artifacts, but once it goes over there would be some minor stencil artifacts if two models with the same stencil id happen to overlap. I think this is a rare situation and a reasonable trade-off. 

Agreed, can you please concisely document in, e.g., if more than 256 models have silhouettes enabled, there is a small chance that overlapping models will have minor artifacts.

The is more a note to us than end users, but it is probably better in the reference doc than as a comment.

**Comment by pjcozzi** [comment]
Do we clear stencil after the `GROUND` pass?

Also, how will this play when we are soon able to render models / 3D Tiles tilesets in the `GROUND` pass?

File: Source/Renderer/Context.js [hunk 25]
```diff
@@ -508,6 +509,18 @@ define([
...
 
         /**
+         * <code>true</code> if the WebGL context supports stencil buffers.
+         * Stencil buffers are not supported by all systems.
+         * @memberof Context.prototype
+         * @type {Boolean}
+         */
+        stencilBuffer : {
+            get : function() {
+                return this._stencilBits >= 8;
```
Comment (pjcozzi) [inline Source/Renderer/Context.js]: I've never seen less than 8 (or anything other than 8), but are you sure this is the right test?

File: Source/Renderer/RenderState.js [hunk 26]
```diff
@@ -589,7 +589,7 @@ define([
             // Section 6.8 of the WebGL spec requires the reference and masks to be the same for
             // front- and back-face tests.  This call prevents invalid operation errors when calling
             // stencilFuncSeparate on Firefox.  Perhaps they should delay validation to avoid requiring this.
-            gl.stencilFunc(stencilTest.frontFunction, stencilTest.reference, stencilTest.mask);
+            gl.stencilFunc(frontFunction, reference, mask);
```
Comment (pjcozzi) [inline Source/Renderer/RenderState.js]: Good eye.

File: Source/Scene/Model.js [hunk 27]
```diff
@@ -3524,6 +3588,227 @@ define([
...
+    function isInvisible(model) {
+        return (model.color.alpha === 0.0);
+    }
+
+    function alphaDirty(currAlpha, prevAlpha) {
+        // Returns whether the alpha state has changed between invisible, translucent, or opaque
+        return (Math.floor(currAlpha) !== Math.floor(prevAlpha)) || (Math.ceil(currAlpha) !== Math.ceil(prevAlpha));
+    }
+
+    // TODO : alternatively increment silhouette length as part of frame state and edit the model render states on the fly
```
Comment (pjcozzi) [inline Source/Scene/Model.js]: What's the plan here?

If we are able to use derived commands to easily add silhouettes to other primitives, we could move this; otherwise, I think it is OK here.

**Comment by pjcozzi** [comment]
@lilleyse can you move the tasklist, https://github.com/AnalyticalGraphicsInc/cesium/pull/4314#issuecomment-265608700, to the top comment (so it is easier to find) and update it?

**Comment by pjcozzi** [comment]
> A possible solution here is to create a derived command that combines multiple commands and always renders them together. All the model commands would be rendered sequentially, followed by all the stencil commands. Does this approach sound ok?

Yes, it does.

File: Source/Renderer/Context.js (hunk 25 above)
Comment (lilleyse) [inline Source/Renderer/Context.js]: When the context is created without `stencil : true`, `gl.getParameter(gl.STENCIL_BITS);` returns 0 so this test should be fine.

**Comment by lilleyse** [comment]
> Also, how will this play when we are soon able to render models / 3D Tiles tilesets in the GROUND pass?

Just to be sure, do you mean the GLOBE pass? Either way it will be problematic...

File: Source/Renderer/Context.js (hunk 25 above)
Comment (pjcozzi) [inline Source/Renderer/Context.js]: I get that it is fine in practice (we think), but are you sure that `> 0` isn't more correct according to the spec?

**Comment by pjcozzi** [comment]
>> Also, how will this play when we are soon able to render models / 3D Tiles tilesets in the GROUND pass?

> Just to be sure, do you mean the GLOBE pass? Either way it will be problematic...

Yes, GLOBE.

> Edit: Though if the silhouettes are opaque then it is doable to clear the stencil after the GLOBE pass. Translucent silhouettes won't be easily possible.

Couldn't the GLOBE pass have OPAQUE and TRANSLUCENT passes with proper stencil clearing?  Or is OIT the issue?

**Comment by lilleyse** [comment]
> Couldn't the GLOBE pass have OPAQUE and TRANSLUCENT passes with proper stencil clearing? Or is OIT the issue?

Ok yeah, if the globe pass has opaque and translucent passes in it this should all be fine.

**Comment by lilleyse** [comment]
I should have this ready later tonight after testing on my other machine.

**Comment by lilleyse** [comment]
The tests are in. I'm still figuring out how to improve the silhouettes in 2D.

**Comment by pjcozzi** [comment]
> I am actually surprised that this works at all since getContext does not create a stencil buffer by default

Mystery solved by @lilleyse: #4747

File: Source/Renderer/Context.js (hunk 25 above)
Comment (pjcozzi) [inline Source/Renderer/Context.js]: OK, thanks for checking.

**Comment by pjcozzi** [comment]
@lilleyse let merge this.  If silhouettes in 2D/CV get widely used, we can improve them.

Likewise, this can also use derived commands when ready.

**Comment by pjcozzi** [comment]
Thanks again for kicking this off, @jasonbeverage.  BIG contribution!

**Comment by jasonbeverage** [comment]
Thanks for merging this @pjcozzi and all the cleanup @lilleyse.  Happy to be able to contribute to Cesium!

**Comment by pjcozzi** [comment]
@shunter do you, by chance, want to do the CZML updates here?
